### PR TITLE
Support Jenkins JNLP agents on Windows-based Docker infrastructure

### PR DESCRIPTION
**Pull request by pronovic**
This pull request adds functionality to support launching JNLP agents on Windows-based Docker infrastructure.   See discussion in issue #151.

The pull request also adds `.gitattributes` configuration that makes it possible to successfully build the plugin on Windows.

The Windows JNLP implementation was based on [pull request #465](https://github.com/jenkinsci/docker-plugin/pull/465) for the [jenkinsci/docker-plugin](https://github.com/jenkinsci/docker-plugin) project.   That pull request originated at [nanori/docker-plugin](https://github.com/nanori/docker-plugin).    The patch no longer applies cleanly, so I used it as a starting point and implemented equivalent functionality that works for [KostyaSha/yet-another-docker-plugin](https://github.com/KostyaSha/yet-another-docker-plugin).

The new implementation was tested by me and @louisn on Windows 10 Enterprise with Docker 17.03.1-ce-win12 (12058).  We also confirmed that existing JNLP functionality for Linux (RHEL 7, Docker 1.2.16) continues to work as it did previously.

All of the existing unit tests pass.   There are not currently any unit tests tied to the methods that I modified, and I did not add any.

Note that pull request #157 is also related to this same functionality.   I did not run across this pull request until after @louisn and I had finished building and testing our version, so I have decided to submit this anyway.  I'm not sure what state that existing pull request is in, since it doesn't seem to have moved recently.

### CODE REVIEW

**Comment by pronovic** [comment]
I should also note that if you ultimately decode to go with #157 instead of pull request, I can submit a separate pull request for the other minor changes, tied to [line endings](https://github.com/pronovic/yet-another-docker-plugin/pull/1) and [ignore files](https://github.com/pronovic/yet-another-docker-plugin/pull/2).

File: .gitignore [hunk 1]
```diff
@@ -6,6 +6,15 @@
...
+.swp
 /target
 
+yet-another-docker-java/dependency-reduced-pom.xml
+yet-another-docker-java/.classpath
+yet-another-docker-java/.project
+yet-another-docker-plugin/.classpath
+yet-another-docker-plugin/.project
+yet-another-docker-plugin/.settings/org.eclipse.jdt.core.prefs
+yet-another-docker-plugin/.settings/de.loskutov.anyedit.AnyEditTools.prefs
```
Comment (KostyaSha) [inline .gitignore]: is it eclipse ignore rules? maybe ignore with `*/.classpath` ? 
https://github.com/github/gitignore/commit/0674ff5b5c77935cbb2ea2269c001bfd84aa1aa5 hm..

File: mvnw.cmd [hunk 2]
```diff
@@ -1,145 +1,145 @@
-@REM ----------------------------------------------------------------------------
```
Comment (KostyaSha) [inline mvnw.cmd]: what was wrong with .cmd? line endings?

Comment (pronovic) [inline mvnw.cmd]: Yes, that commit is the side-effect of adding `.gitattributes`.  It looked like `mvnw.cmd` should have DOS line endings, so I marked it that way in `.gitattributes` and git converted it.  I could revert this change if that wasn't the right decision.

Comment (KostyaSha) [inline mvnw.cmd]: never used windows for development/builds :D last time it ends with workarounds in java specific to windows bugs...

File: .gitignore (hunk 1 above)
Comment (pronovic) [inline .gitignore]: Yes, these rules mostly ignore Eclipse files generated by Maven.  AnyEdit is a plugin I have installed in my workspace.  We could alternately ignore `*/.classpath`, `*/.project`, `*/.settings`, etc.

File: mvnw.cmd (hunk 2 above)
Comment (pronovic) [inline mvnw.cmd]: The code builds fine in Windows, but didn't actually work when installed in Jenkins -- probably because of the wrong line endings in `init.sh`.  When I rebuilt the code Linux and it did work in Jenkins, I decided it must be related to line endings.  Adding `.gitattributes` fixed the problem.

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java [hunk 3]
```diff
@@ -174,26 +175,54 @@ public void launch(@Nonnull SlaveComputer computer, TaskListener listener) throw
...
-                    .withAttachStdout(true)
-                    .withCmd("/bin/sh", "-cxe", startCmd.replace("$", "\\$"))
-                    .exec();
+            String startCmd = null;
+            ExecCreateCmdResponse response = null;
+
+            if (osType.equals("windows")) {
+                startCmd =
+                        "@\"\n" +
+                                "`$JENKINS_URL='"     + rootUrl + "'\n" +
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java]: please add WNL (windows new line) static string.

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java [hunk 4]
```diff
@@ -287,21 +316,36 @@ public ComputerLauncher getLauncher() {
     }
 
     @Override
-    public void appendContainerConfig(DockerSlaveTemplate dockerSlaveTemplate, CreateContainerCmd createContainerCmd)
+    public void appendContainerConfig(DockerSlaveTemplate dockerSlaveTemplate, CreateContainerCmd createContainerCmd, String osType)
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java]: i feel that will be not the last variable... 
Could you make some `AdditionalInfo` class and move `osType` to it?

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java (hunk 3 above)
Comment (pronovic) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java]: ok.

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java (hunk 4 above)
Comment (pronovic) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java]: yes, no problem.

**Comment by pronovic** [comment]
I will work on changes and submit an update to the pull request tomorrow.   If there is anything else you would like to see changed, please let me know.

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java [hunk 5]
```diff
@@ -221,12 +221,13 @@ private DockerSlave provisionWithWait(DockerSlaveTemplate template, Provisioning
         final DockerContainerLifecycle dockerContainerLifecycle = template.getDockerContainerLifecycle();
         final String imageId = dockerContainerLifecycle.getImage();
         final DockerComputerLauncher computerLauncher = template.getLauncher();
+        final String osType = this.getClient().infoCmd().exec().getOsType();
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: please don't use `this` when there is no overlapping methods.

**Comment by KostyaSha** [comment]
Just to describe this spaghetty.
DockerTemplate is configuration that saved from UI and used as base for running different slaves. Launcher has internal state variable specific for single live slaves, so launcherconfiguration is `cloned` with `getPreparedLauncher()` into separate object that used then in Slave lifecycle.
We can't store docker daemon specific states in it, but it the only available argument. So ostype was added.
But shouldn't be image type specific to image? Shouldn't we check Image info instead daemon info? 
Then we can have transient synchronised field with osType, ideally enum. And once we resolve that this image is windows - keep it. On next configuration from UI or jenkins restart transient will be reseted.

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java [hunk 6]
```diff
@@ -165,6 +165,7 @@ public void launch(@Nonnull SlaveComputer computer, TaskListener listener) throw
         }
         final DockerSlaveTemplate dockerSlaveTemplate = node.getDockerSlaveTemplate();
         final DockerComputerJNLPLauncher launcher = (DockerComputerJNLPLauncher) dockerSlaveTemplate.getLauncher();
+        final String osType = connect.infoCmd().exec().getOsType();
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java]: see comment

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java (hunk 5 above)
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: see comment. So here you have template and client

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java [hunk 7]
```diff
@@ -221,12 +221,13 @@ private DockerSlave provisionWithWait(DockerSlaveTemplate template, Provisioning
...
 
         //pull image
         dockerContainerLifecycle.getPullImage().exec(getClient(), imageId);
 
         LOG.info("Trying to run container for {}", imageId);
-        final String containerId = runContainer(template);
+        final String containerId = runContainer(template, osType);
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: after image pull you can get info about image and store it in template. AFAIK image os type is specific to image, not to daemon.

Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: https://docs.docker.com/engine/api/v1.26/#operation/ImageInspect

**Comment by pronovic** [comment]
Ok, I think I understand what you are looking for.  I will play with some refactoring and submit and update the pull request when I have something solid.   It will probably be sometime next week instead of tomorrow.

**Comment by KostyaSha** [comment]
I think code will become more accurate, additional argument will disappear

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerSSHLauncher.java [hunk 8]
```diff
@@ -63,18 +63,23 @@ public ComputerLauncher getPreparedLauncher(String cloudId, DockerSlaveTemplate
...
-        createCmd.withPortSpecs(sshPort + "/tcp");
-
-        String[] cmd = dockerSlaveTemplate.getDockerContainerLifecycle().getCreateContainer().getDockerCommandArray();
-        if (cmd.length == 0) {
-            //default value to preserve compatibility
-            createCmd.withCmd("bash", "-c", "/usr/sbin/sshd -D -p " + sshPort);
+    public void appendContainerConfig(DockerSlaveTemplate dockerSlaveTemplate, CreateContainerCmd createCmd, String osType) {
+        if (osType.equals("windows")) {
+            // TODO: figure out how to support SSH launcher on Windows
+            throw new RuntimeException("Windows OS is not supported for SSH launcher");
```
Comment (gamma) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerSSHLauncher.java]: I would not do this. This will even make the usage of cygwin sshd impossible. We're using it quite successfully.

Comment (gamma) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerSSHLauncher.java]: In addition: I had it almost running with the win-openssl implementation which lacks a fix (dunno which exactly).

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java (hunk 5 above)
Comment (gamma) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: Have you checked a swarm? AFAIK the swarm will not return an osType. I did not know of the osType of an Image that can be obtained by inspecting it ... so that should be the way to go.

**Comment by oleg-nenashev** [comment]
FYI there is Work Directory support in Remoting, which will be added to Jenkins Weeklys soon: https://issues.jenkins-ci.org/browse/JENKINS-44108

File: yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1 [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+$RUN_CMD="$RUN_CMD -jnlpUrl $JENKINS_URL/$COMPUTER_URL/slave-agent.jnlp"
+
+if ($COMPUTER_SECRET) {
+   $RUN_CMD="$RUN_CMD -secret $COMPUTER_SECRET"
+}
+
+cd $JENKINS_HOME
+
+try {
+   Write-Output "Invoke-WebRequest -Uri $JENKINS_URL/jnlpJars/slave.jar -Outfile $JENKINS_HOME/slave.jar"
```
Comment (oleg-nenashev) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1]: See the discussion in https://github.com/jenkinsci/windows-slave-installer-module/pull/6 . There we agreed that doing such automatic upload via HTTP is not a safe approach.  Maybe it worth sending resource to the agent with Docker tools before invoking PowerShell

File: yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1 [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+if ($NO_CERTIFICATE_CHECK -eq "true") {
+   $RUN_CMD="$RUN_CMD -noCertificateCheck"
+}
+
+if (!$SLAVE_OPTS) {
+   $RUN_CMD="$RUN_CMD $SLAVE_OPTS"
+}
+
+$RUN_CMD="$RUN_CMD -jnlpUrl $JENKINS_URL/$COMPUTER_URL/slave-agent.jnlp"
```
Comment (oleg-nenashev) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1]: With such configuration there will be no logging to the disk by default. Probably it is fine for Docker agent (STDOUT/STDERR will be in `docker logs` anyway)

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerSSHLauncher.java (hunk 8 above)
Comment (oleg-nenashev) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerSSHLauncher.java]: Agreed

File: yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1 (hunk 9 above)
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1]: I will replaced with docker copy soon. but it pretty safe because jnlp connection do the same as slave.jar fetch.

File: yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1 (hunk 10 above)
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1]: i added in some placed docker log print, but without better cloud stats support https://issues.jenkins-ci.org/browse/JENKINS-43788 it impossible

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java (hunk 5 above)
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: 99% expect ostype to come from image, that will be correct and work without adding additional arguments.

File: yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1 (hunk 9 above)
Comment (oleg-nenashev) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1]: agreed

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerSSHLauncher.java (hunk 8 above)
Comment (pronovic) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerSSHLauncher.java]: I can remove this check.

File: yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1 (hunk 9 above)
Comment (pronovic) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1]: In this case, the Powershell script just reproduces the behavior in the existing `init.sh` script.   Is it ok to leave this as-is for now?

Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1]: yes, sure, dunno when i will have time to change something already working :D

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java (hunk 5 above)
Comment (pronovic) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: I have not checked Swarm.   Basically, all I did was port the [existing pull request for docker-plugin](https://github.com/jenkinsci/docker-plugin/pull/465) into this codebase and test the result in my infrastructure.  I agree, it would be better to get the OS type from the image.  That way, it won't matter where the image is running.

File: yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1 (hunk 10 above)
Comment (pronovic) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1]: In this case, the Powershell script just reproduces the behavior in the existing `init.sh` script. Is it ok to leave this as-is for now?

File: yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1 (hunk 9 above)
Comment (gamma) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1]: I think a docker copy would help a problem I had with the jnlp not being copied to the slave when connecting to a swarm based slave. (We're hiding the jerkinses behind a proxy)

**Comment by pronovic** [comment]
I have made changes in this pull request.  The following changes are included:

* cleanup in `.gitignore`
* refactor out a `WNL` (windows newline) constant in `DockerComputerJNLPLauncher.java`
* add a new enumeration `DockerOperatingSystem` 
* remove code that blocked the SSH connection type on Windows platform
* derive operating system from the Docker image rather than the runtime container
* save operating system as a transient variable on `DockerSlaveTemplate` 
* remove the `osType` argument passed to various methods and use `DockerSlaveTemplate.getOperatingSystem()` instead

These have been tested with my existing Linux and Windows Docker infrastructure.  I have not tested with Swarm.

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java [hunk 11]
```diff
@@ -225,7 +226,13 @@ private DockerSlave provisionWithWait(DockerSlaveTemplate template, Provisioning
         //pull image
         dockerContainerLifecycle.getPullImage().exec(getClient(), imageId);
 
-        LOG.info("Trying to run container for {}", imageId);
+        // set the operating system if it's not already cached
+        if (template.getOperatingSystem() == null) {
+            DockerOperatingSystem operatingSystem = determineOperatingSystem(imageId);
+            template.setOperatingSystem(operatingSystem);
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: one line call

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java [hunk 12]
```diff
@@ -237,6 +244,8 @@ private DockerSlave provisionWithWait(DockerSlaveTemplate template, Provisioning
             throw ex;
         }
 
+
+
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: remove lines

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java [hunk 13]
```diff
@@ -258,6 +267,21 @@ private DockerSlave provisionWithWait(DockerSlaveTemplate template, Provisioning
...
+    private DockerOperatingSystem determineOperatingSystem(String imageId) {
+        InspectImageResponse ir = getClient().inspectImageCmd(imageId).exec();
+        if ("linux".equals(ir.getOs())) {
+            LOG.info("Detected LINUX operating system for image: {}", imageId);
+            return DockerOperatingSystem.LINUX;
+        } else if ("windows".equals(ir.getOs())) {
+            LOG.info("Detected WINDOWS operating system for image: {}", imageId);
+            return DockerOperatingSystem.WINDOWS;
+        } else {
+            LOG.info("Detected OTHER operating system ({}) for image: {}", ir.getOs(), imageId);
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: it's useless for info, move to TRACE

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerOperatingSystem.java [hunk 14]
```diff
@@ -0,0 +1,8 @@
+package com.github.kostyasha.yad;
+
+/** Supported operating systems. */
+public enum DockerOperatingSystem {
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerOperatingSystem.java]: `OsType`

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerSlaveTemplate.java [hunk 15]
```diff
@@ -36,6 +36,7 @@
     private int maxCapacity = 10;
 
     private transient /*almost final*/ Set<LabelAtom> labelSet;
+    private transient DockerOperatingSystem operatingSystem;
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerSlaveTemplate.java]: `osType`

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java [hunk 16]
```diff
@@ -174,26 +176,54 @@ public void launch(@Nonnull SlaveComputer computer, TaskListener listener) throw
...
-                    .withTty(true)
-                    .withAttachStdin(false)
-                    .withAttachStderr(true)
-                    .withAttachStdout(true)
-                    .withCmd("/bin/sh", "-cxe", startCmd.replace("$", "\\$"))
-                    .exec();
+            String startCmd = null;
+            ExecCreateCmdResponse response = null;
+
+            if (DockerOperatingSystem.WINDOWS.equals(dockerSlaveTemplate.getOperatingSystem())) {
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java]: `==` for enum

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java [hunk 17]
```diff
@@ -289,19 +319,34 @@ public ComputerLauncher getLauncher() {
...
+                    throw new IllegalStateException("Resource file 'init.ps1' not found");
+                }
+
+                createContainerCmd.withEntrypoint("powershell.exe")
+                        .withCmd("powershell.exe", "-NoLogo", "-ExecutionPolicy", "bypass", "-Command", "{@\"\n" +
+                        initCmd.replace("$", "`$") +
+                        "\n\"@ | Out-File -FilePath c:\\init.ps1 ; if($?) {c:\\init.ps1}}");
+            }
+        } else { // default is DockerOperatingSystem.LINUX
+            try (InputStream instream = DockerComputerJNLPLauncher.class.getResourceAsStream("DockerComputerJNLPLauncher/init.sh")) {
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher.java]: `inStream`

**Comment by pronovic** [comment]
Comments addressed.

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java [hunk 18]
```diff
@@ -258,6 +264,21 @@ private DockerSlave provisionWithWait(DockerSlaveTemplate template, Provisioning
         return new DockerSlave(slaveName, nodeDescription, launcher, containerId, template, getDisplayName(), id);
     }
 
+    /** Determine the operating system associated with an image. */
+    private OsType determineOsType(String imageId) {
+        InspectImageResponse ir = getClient().inspectImageCmd(imageId).exec();
+        if ("linux".equals(ir.getOs())) {
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: isn't it `Linux` ? better set `equalsIgnoreCase`

File: yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1 (hunk 10 above)
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.ps1]: yes

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java (hunk 18 above)
Comment (pronovic) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: On my test systems, I see `"windows"` and `"linux"`.  However, I can change it to `equalsIgnoreCase` to be safe.

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java [hunk 19]
```diff
@@ -225,7 +226,12 @@ private DockerSlave provisionWithWait(DockerSlaveTemplate template, Provisioning
         //pull image
         dockerContainerLifecycle.getPullImage().exec(getClient(), imageId);
 
-        LOG.info("Trying to run container for {}", imageId);
+        // set the operating system if it's not already cached
+        if (template.getOsType() == null) {
+            template.setOsType(determineOsType(imageId));
+        }
+
+        LOG.info("Trying to run container for {}, operating system {}", imageId, template.getOsType());
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerCloud.java]: info->debug

**Comment by pronovic** [comment]
Just wanted to check in and see whether any other changes are required?  I haven't seen any other review comments.

**Comment by KostyaSha** [comment]
Please squash into one commit to not lost gh contributions for you

**Comment by pronovic** [comment]
Ok, it's squashed.

**Comment by KostyaSha** [comment]
@gamma @oleg-nenashev also feel free to review, but i see no issues and merging it.

**Comment by jlouisfoster** [comment]
Hi,

Is there any documentation for how to properly setup a windows container using YADP - specifically in relation to this fix?
